sarif/src/main/java/com/jetbrains/qodana/sarif/baseline/DescriptorLookup.java

```diff
@@ -0,0 +1,47 @@
+package com.jetbrains.qodana.sarif.baseline;
+
+import com.jetbrains.qodana.sarif.model.ReportingDescriptor;
+import com.jetbrains.qodana.sarif.model.Run;
+import com.jetbrains.qodana.sarif.model.Tool;
+
+import java.util.*;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+final class DescriptorLookup {
+    private final Map<String, Optional<DescriptorWithLocation>> lookup;
+    private final Supplier<Stream<DescriptorWithLocation>> rules;
+
+    DescriptorLookup(Run run) {
+        this.lookup = new HashMap<>();
+        this.rules = () -> descriptors(run);
+    }
+
+    Optional<DescriptorWithLocation> findById(String ruleId) {
+        return lookup.computeIfAbsent(ruleId, id ->
+                rules.get().filter(r -> Objects.equals(r.getDescriptor().getId(), id)).findFirst()
+        );
+    }
+
+    private Stream<DescriptorWithLocation> descriptors(Run run) {
```

Comment:
This lazy-but-repeated approach is taken from the converter. Saves us creating a huge map if we only ever need to check for very few IDs - on the other hand it's wasteful if there are a lot of different IDs to be checked.

Chose this, because I think the "good" case is more likely here.